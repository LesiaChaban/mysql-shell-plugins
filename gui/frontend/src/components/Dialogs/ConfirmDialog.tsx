/*
 * Copyright (c) 2021, 2022, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import "./ConfirmDialog.css";

import React from "react";

import {
    Button, Codicon, Component, Container, Dialog, IComponentProperties, IComponentState, Icon, Label, Orientation,
} from "../ui";

export interface IConfirmDialogProperties extends IComponentProperties {
    caption: string;
    onClose?: (accepted: boolean, payload?: unknown) => void;
}

export interface IConfirmDialogState extends IComponentState {
    message: React.ReactNode;
    refuseText: string;
    acceptText: string;
    payload?: unknown;
}

export class ConfirmDialog extends Component<IConfirmDialogProperties, IConfirmDialogState> {

    private dialogRef = React.createRef<Dialog>();

    public constructor(props: IConfirmDialogProperties) {
        super(props);
        this.state = {
            message: "",
            refuseText: "",
            acceptText: "",
        };
        this.addHandledProperties("caption", "onClose");
    }

    public show = (message: React.ReactNode, refuseText: string, acceptText: string, payload?: unknown): void => {
        this.setState({ message, refuseText, acceptText, payload }, () => {
            return this.dialogRef.current?.open();
        });
    };

    public render(): React.ReactNode {
        const { caption } = this.props;
        const { message, refuseText, acceptText } = this.state;

        const className = this.getEffectiveClassNames(["confirmDialog"]);
        let dialogContent = null;
        if (React.isValidElement(message)) {
            dialogContent = message;
        } else {
            dialogContent =
                <Container orientation={Orientation.TopDown}>
                    {message && <Label id="dialogMessage" caption={message as string} />}
                </Container>;
        }

        const actions: React.ReactNode[] = [];
        if (acceptText) {
            actions.push(<Button
                caption={acceptText}
                id="accept"
                key="accept"
                onClick={this.handleActionClick}
            />);
        }
        if (refuseText) {
            actions.push(<Button
                caption={refuseText}
                id="refuse"
                key="refuse"
                onClick={this.handleActionClick}
            />);
        }

        return (
            <Dialog
                ref={this.dialogRef}
                className={className}
                caption={
                    <>
                        <Icon src={Codicon.Question} />
                        <Label>{caption}</Label>
                    </>
                }
                content={dialogContent}
                actions={{
                    end: actions,
                }}
            >
            </Dialog>
        );
    }

    private handleActionClick = (e: React.SyntheticEvent, props: Readonly<IComponentProperties>): void => {
        const { onClose } = this.props;
        const { payload } = this.state;

        let accepted = false;

        if (props.id === "accept") {
            accepted = true;
        }

        this.dialogRef.current?.close(!accepted);

        onClose?.(accepted, payload);
    };

}
