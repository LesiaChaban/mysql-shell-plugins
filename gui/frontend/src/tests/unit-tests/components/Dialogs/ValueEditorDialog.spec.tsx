/*
 * Copyright (c) 2022, Oracle and/or its affiliates.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License, version 2.0,
 * as published by the Free Software Foundation.
 *
 * This program is also distributed with certain software (including
 * but not limited to OpenSSL) that is licensed under separate terms, as
 * designated in a particular file or component or in included license
 * documentation.  The authors of MySQL hereby grant you an additional
 * permission to link the program and your derivative works with the
 * separately licensed software that they have included with MySQL.
 * This program is distributed in the hope that it will be useful,  but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
 * the GNU General Public License, version 2.0, for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA
 */

import { render } from "@testing-library/preact";
import { mount } from "enzyme";
import React from "react";
import { IDialogSection, ValueEditDialog } from "../../../../components/Dialogs/ValueEditDialog";

const dialogRef = React.createRef<ValueEditDialog>();

describe("Value Editor Dialog output tests", (): void => {
    it("Action output elements", () => {
        const component = mount<ValueEditDialog>(
            <ValueEditDialog
                ref={dialogRef}
                caption="Enter a name for the new theme"
                onValidate={jest.fn()}
                onClose={jest.fn()}
            />,
        );
        const instance = component.instance();
        const nameSection: IDialogSection = {
            values: {
                input: {
                    caption: "input",
                    value: "",
                    span: 8,
                },
            },
        };

        instance.show(
            {
                id: "testDialog",
                sections: new Map<string, IDialogSection>([
                    ["name", nameSection],
                ]),
            },
            [],
        );
        // const spyOnValidate = jest.spyOn(instance.props as IValueEditDialogProperties, "onValidate");
        // const spyOnOk = jest.spyOn(instance.props as IValueEditDialogProperties, "onClose");
        expect(component).toBeTruthy();
    });

    it("Value Editor Dialog (Snapshot) 1", () => {
        const component = render(
            <ValueEditDialog
                ref={dialogRef}
                caption="Enter a name for the new theme"
                onValidate={jest.fn()}
                onClose={jest.fn()}
            />,
        );
        expect(component).toMatchSnapshot();
    });

});
