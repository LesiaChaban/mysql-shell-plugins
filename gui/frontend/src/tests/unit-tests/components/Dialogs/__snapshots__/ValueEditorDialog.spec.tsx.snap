// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Value Editor Dialog output tests Value Editor Dialog (Snapshot) 1 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body
    theme="dark"
  >
    <div
      __self="[object Object]"
      __source="[object Object]"
      class="msg portal"
      id="testDialog"
    >
      <div
        __self="[object Object]"
        __source="[object Object]"
        class="msg dialog visible valueEditDialog"
      >
        <div
          __self="[object Object]"
          __source="[object Object]"
          class="title verticalCenterContent"
        >
          <div
            __self="[object Object]"
            __source="[object Object]"
            class="msg icon codicon codicon-terminal"
          />
          <label
            __self="[object Object]"
            __source="[object Object]"
            class="msg label"
            data-tooltip="expand"
          >
            Enter a name for the new theme
          </label>
        </div>
        <div
          __self="[object Object]"
          __source="[object Object]"
          class="content fixedScrollbar"
        >
          <div
            __self="[object Object]"
            __source="[object Object]"
            class="msg grid"
            style="grid-template-columns: repeat(8, 1fr); row-gap: 16px; column-gap: 16px;"
          >
            <div
              __self="[object Object]"
              __source="[object Object]"
              class="msg container fixedScrollbar gridCell"
              style="flex-direction: column; justify-content: flex-start; align-items: stretch; grid-column: span 8;"
            >
              <label
                __self="[object Object]"
                __source="[object Object]"
                class="msg label valueTitle"
                data-tooltip="expand"
              >
                input
              </label>
              <div
                __self="[object Object]"
                __source="[object Object]"
                class="msg container fixedScrollbar inputWithProgress"
                style="flex-direction: row;"
              >
                <input
                  __self="[object Object]"
                  __source="[object Object]"
                  class="msg input valueEditor"
                  id="input"
                  spellcheck="true"
                  type="text"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          __self="[object Object]"
          __source="[object Object]"
          class="footer verticalCenterContent unselectable"
        >
          <div
            __self="[object Object]"
            __source="[object Object]"
            class="msg container fixedScrollbar leftItems"
            style="flex-direction: row;"
          />
          <div
            __self="[object Object]"
            __source="[object Object]"
            class="msg container fixedScrollbar rightItems"
            style="flex-direction: row-reverse;"
          >
            <button
              __self="[object Object]"
              __source="[object Object]"
              caption="OK"
              class="msg button"
              id="ok"
            >
              OK
            </button>
            <button
              __self="[object Object]"
              __source="[object Object]"
              caption="Cancel"
              class="msg button"
              id="cancel"
            >
              Cancel
            </button>
          </div>
        </div>
      </div>
    </div>
    <div />
  </body>,
  "container": <div />,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
